### PR TITLE
lock before accessing class variables in stop

### PR DESCRIPTION
**Pull request by mrkraimer**
First let me thank Michael for weak pointer fixes.
While looking for the problem I did see one other bug.
In all methods of caContext.cpp except stop 
Lock xx(mutex);
was called before any class variables were accessed. 
It should also be done in the stop method.

When this was done before Michael's changes it always caused a server crash when a monitor client exited.

I have tested with both 3.14 and 3.15 and do not see any more crashes.

### CODE REVIEW

**Comment by ralphlange** [comment]
I don't really understand your comment. Three questions:

1. Doing what caused crashes...
2. ...before which of Michael's changes?
3. What did you test when seeing no more crashes? This proposal, I assume, but with some changes of Michael or without?

**Comment by mrkraimer** [comment]
On 03/17/2016 07:35 AM, Ralph Lange wrote:
>
> I don't really understand your comment. Three questions:
>
>  1. Doing what caused crashes...
>  2. ...before which of Michael's changes?
>  3. What did you test when seeing no more crashes? This proposal, I
>     assume, but with some changes of Michael or without?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/epics-base/pvaSrv/pull/35#issuecomment-197838182>
>

Monitors are the only  Channel methods that can be not implemented by 
using only dbAccess methods.
Because db_post_event is implemented by channel access (rsrv through at 
least the 3.15 releases) this means
that ca client calls must be used to implement monitors.

In order to use ca properly the implementation must provide a thread to 
receive the monitors issued as a result
of db_post_event calls.

The code in caContext.h and caContext.cpp is the code that supplies the 
thread.
It also allows the thread to be shared by all  montor requests that come 
from the same pvAccess thread,
i. e. that come from the same pvAccess client thread.

The caContext thread is created when the pvAccess client thread creates 
the first monitor.
Additional monitors created by the same client thread will share the 
same caContext thread.
When the last monitor from that client is destroyed then the caContext 
thread is also destroyed.

The code that can cause the caContext thread to be destroyed is in

void caContext::stop()
{
     if(DbPvDebug::getLevel()>0) printf("caContext::stop\n");
     epicsThreadId id = epicsThreadGetIdSelf();
     if(id!=threadId) {
         printf("caContext::stop not same thread\n");
         return;
         //throw std::logic_error(String(
         //   "caContext::stop not same thread"));
     }
     Lock xx(mutex);  // NOTE NOTE THIS IS WHAT I AM ADDING
     if(referenceCount!=0) {
         printf("caContext::stop referenceCount != 0 value %d\n",
             referenceCount);
         return;
         //throw std::logic_error(String(
         //   "caContext::stop referenceCount != 0"));
     }
     else
     {
         caContextCreate::erase(threadId);
         ca_context_destroy();
     }
}

The call to caContextCreate::erase(threadId) is what causes the 
caContext thread to exit,
when the referenceCount goes to 0.
As a side effect of the way caContextCreate::erase works then before 
weak pointers were added
there was no longer any code that had a shared_pointer to the caContext 
instance.
Thus C++ deleted it.
Thus the mutex in Lock xx(mutex) no longer exists.
So when the lock is unlocked when caContext::stop() returns a crash occurs.

Michael's code changed

static void threadExitFunc(void *arg)
{
     if(DbPvDebug::getLevel()>0) printf("caContext::threadExitFunc\n");
     caContext * context = static_cast<caContext * >(arg);
     context ->stop();
}

to

static void threadExitFunc(void *arg)
{
     if(DbPvDebug::getLevel()>0) printf("caContext::threadExitFunc\n");
     caContext * context = static_cast<caContext * >(arg);
     std::tr1::shared_ptr<caContext> self(context->shared_from_this());
     self->stop();
}

This there was still a shared_pointer to caContext until threadExitFunc 
returns.

Thus all OK.

Hope this helps.

Note that if future versions of base implement db_post_event in dbAccess 
(or it's successor) instead of rsrv,
Then the pvaSrv code for monitors may become less complex.

**Comment by mrkraimer** [comment]
On 03/17/2016 07:35 AM, Ralph Lange wrote:
>
> I don't really understand your comment. Three questions:
>
>  1. Doing what caused crashes...
>  2. ...before which of Michael's changes?
>  3. What did you test when seeing no more crashes? This proposal, I
>     assume, but with some changes of Michael or without?
>

Sorry I did not answer 3)..

Before I made the one line code change to add the lock I was up to date 
with latest master branch from epics-base.
This was after you merged Michael's changes.

Marty

**Comment by mdavidsaver** [comment]
@mrkraimer I'm afraid I don't understand your explanation.  From the code, I'm guessing that you're worried about potentially concurrent access to ```referenceCount```,which seems a legitimate.  Continuing to hold the lock for ```caContextCreate::erase(threadId)``` and ```ca_context_destroy()``` is unnecessary, but I think harmless.

> Note that if future versions of base implement db_post_event in dbAccess (or it's successor) instead of rsrv, Then the pvaSrv code for monitors may become less complex.

As I discovered last fall, it is possible to use the dbEvent.h API directly to get event callbacks, just as RSRV does.  So there is no need to use libca at all.

**Comment by mrkraimer** [comment]
On 03/17/2016 02:16 PM, mdavidsaver wrote:
>
> @mrkraimer <https://github.com/mrkraimer> I'm afraid I don't 
> understand your explanation. From the code, I'm guessing that you're 
> worried about potentially concurrent access to |referenceCount|,which 
> seems a legitimate. Continuing to hold the lock for 
> |caContextCreate::erase(threadId)| and |ca_context_destroy()| is 
> unnecessary, but I think harmless.
>
>     Note that if future versions of base implement db_post_event in
>     dbAccess (or it's successor) instead of rsrv, Then the pvaSrv code
>     for monitors may become less complex.
>
> As I discovered last fall, it is possible to use the dbEvent.h API 
> directly to get event callbacks, just as RSRV does. So there is no 
> need to use libca at all.
>

Looks interesting.
If I understand correctly than  dbEvent handles threading so no need for 
caContext.

Since You and Ralph have the best understanding of both pvAccess AND 
future plans for epics_base than
one of You is best prepared to make the change to implement monitoring 
via dbEvent instead of using libca.

**Comment by ralphlange** [comment]
Sorry, Marty.
I didn't intend to have you create an epic description of the changes and their side effects. I am trying to follow all V4 developments, and in most cases I am able to understand what's happening.

My questions were of simple semantical nature.
You wrote
```
In all methods ... except stop, Lock()
was called before ... 
It should also be done in the stop method.

When this was done before Michael's changes ...
```
When what was done? Calling Lock()? Not calling Lock()? Adding the Lock() call to the stop method? Your sentence is ambiguous in this part.

Which of Michael's changes? Michael is doing a lot of changes in a lot of modules. Even when limiting the scope to his changes about the server crashes handled in #30, there are at least two changes (within hours) that you could have meant: the c801118 workaround in PR #34 commited to pvaSrv, and two changes in Base 3.14 ([12617](http://bazaar.launchpad.net/~epics-core/epics-base/3.14/revision/12617) and [12618](http://bazaar.launchpad.net/~epics-core/epics-base/3.14/revision/12618)) that fix the root cause. This is also ambiguous.

Bottom line:
Please discuss things in the GitHub issues that deal with them.
When mentioning changes, use the short id of the commit (which becomes a link to the diff, see above).
When mentioning people or other issues, use the `@<username>` and `#<issue>` syntax that creates links (see above, buttons in the web interface make this easier).

Emoji, however, are not mandatory. :wink:

**Comment by ralphlange** [comment]
I have created #36 for the switch to the dbEvent API.

**Comment by ralphlange** [comment]
In the end, this is a pull request, not an IRC.
The proposed change seems justified and appropriate, merging.

**Comment by mdavidsaver** [comment]
> Continuing to hold the lock for caContextCreate::erase(threadId) and ca_context_destroy() is unnecessary, but I think harmless.

It occurs to me that my statement is not correct.  Holding this lock while calling ca_context_destroy() introduced a possible deadlock since it is also taken during the exception() callback.  So an deadlock could occur in the unlikely situation of an exception callback just as the context is being destroyed.

**Comment by ralphlange** [comment]
So what do you suggest?

**Comment by mdavidsaver** [comment]
Forgive me if I simply demonstrate 3bbec69f930c41c26ccda11b4f941e069e3b4504

**Comment by mrkraimer** [comment]
On 03/18/2016 10:26 AM, mdavidsaver wrote:
>
> Forgive me if I simply demonstrate 3bbec69 
> <https://github.com/epics-base/pvaSrv/commit/3bbec69f930c41c26ccda11b4f941e069e3b4504>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/epics-base/pvaSrv/pull/35#issuecomment-198384359>
>
I agree with the changes.

Michael,

Do you want to create a new pull request or do You want me to do it?

Marty

**Comment by mdavidsaver** [comment]
@mrkraimer I'm asking for forgiveness because this change is already done and pushed.

**Comment by mrkraimer** [comment]
On 03/18/2016 01:18 PM, mdavidsaver wrote:
>
> @mrkraimer <https://github.com/mrkraimer> I'm asking for forgiveness 
> because this change is already done and pushed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/epics-base/pvaSrv/pull/35#issuecomment-198459775>
>
OK you are forgiven

Marty
